python/ray/monitor.py

```diff
@@ -0,0 +1,187 @@
+    if message is None:
+      return None
+
+    # Parse the message.
+    data = message["data"]
+    db_client_id = data[:DBClientID_SIZE]
+    data = data[DBClientID_SIZE + 1:]
+    data = data.split(b" ")
+    client_type, auxiliary_address, is_insertion = data
+    is_insertion = int(is_insertion)
+    if is_insertion != 1 and is_insertion != 0:
+      raise Exception("Expected 0 or 1 for insertion field, got {} instead".format(is_insertion))
+    is_insertion = bool(is_insertion)
+
+    return db_client_id, client_type, auxiliary_address, is_insertion
+
+  def cleanup_task_table(self):
+    """Clean up global state for a failed local schedulers.
+
+    This marks any tasks that were scheduled on dead local schedulers as
+    TASK_STATUS_LOST. A local scheduler is deemed dead if it is not in
+    self.local_schedulers.
+    """
+    task_ids = self.redis.scan_iter(match="{prefix}*".format(prefix=TASK_PREFIX))
+    for task_id in task_ids:
+      task_id = task_id[len(TASK_PREFIX):]
+      response = self.redis.execute_command("RAY.TASK_TABLE_GET", task_id)
+      if response[1] not in self.local_schedulers:
+        ok = self.redis.execute_command("RAY.TASK_TABLE_UPDATE",
+                                        task_id,
+                                        TASK_STATUS_LOST,
+                                        NIL_ID)
+        if ok != b"OK":
+          log.warn("Failed to update lost task for dead scheduler.")
+
+  def scan_db_client_table(self):
+    """Scan the database client table for the current clients.
+
+    After subscribing to the client table, it's necessary to call this before
+    reading any messages from the subscription channel.
+    """
+    db_client_keys = self.redis.keys("{prefix}*".format(prefix=DB_CLIENT_PREFIX))
+    for db_client_key in db_client_keys:
+      db_client_id = db_client_key[len(DB_CLIENT_PREFIX):]
+      client_type = self.redis.hget(db_client_key, "client_type")
+      if client_type == LOCAL_SCHEDULER_CLIENT_TYPE:
+        self.local_schedulers.add(db_client_id)
+
+  def run(self):
+    """Run the monitor.
+
+    This function loops forever, checking for messages about dead database
+    clients and cleaning up state accordingly.
+    """
+    # Initialize the subscription channel.
+    self.subscribe()
+
+    # Scan the database table and clean up any state associated with clients
+    # not in the database table. NOTE: This must be called before reading any
+    # messages from the subscription channel. This ensures that we start in a
+    # consistent state, since we may have missed notifications that were sent
+    # before we connected to the subscription channel.
+    self.scan_db_client_table()
+    self.cleanup_task_table()
+    log.debug("Scanned schedulers: {}".format(self.local_schedulers))
+
+    # Read messages from the subscription channel.
+    while True:
+      time.sleep(LOCAL_SCHEDULER_HEARTBEAT_TIMEOUT_MILLISECONDS / 1000.)
```

Comment:
we can replace `1000.` with `1000`, since we are doing `from __future__ import division` at the top